tensorstore/kvstore/s3/credentials/file_credential_provider.cc

```diff
@@ -0,0 +1,147 @@
+  std::string result;
+
+  if (auto credentials_file = GetEnv(kEnvAwsCredentialsFile);
+      credentials_file) {
+    result = *credentials_file;
+  } else {
+    if (auto home_dir = GetEnv("HOME"); home_dir) {
+      result = JoinPath(*home_dir, kDefaultAwsCredentialsFilePath);
+    } else {
+      return absl::NotFoundError("Could not read $HOME");
+    }
+  }
+
+  if (auto fstream = std::ifstream(result.c_str()); !fstream.good()) {
+    return absl::NotFoundError(
+        absl::StrCat("Could not find the credentials file at "
+                     "location [",
+                     result, "]"));
+  }
+
+  return result;
+}
+
+}  // namespace
+
+/// https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html#cli-configure-files-format
+Result<AwsCredentials> FileCredentialProvider::GetCredentials() {
+  TENSORSTORE_ASSIGN_OR_RETURN(auto filename, ([&]() -> Result<std::string> {
+                                 if (!filename_.empty()) return filename_;
+                                 return GetAwsCredentialsFileName();
+                               }()));
+
+  std::ifstream ifs(filename.c_str());
+  if (!ifs) {
+    return absl::NotFoundError(
+        absl::StrCat("Could not open credentials file [", filename, "]"));
+  }
+
+  auto profile = !profile_.empty()
+                     ? std::string(profile_)
+                     : GetEnv(kEnvAwsProfile).value_or(kDefaultProfile);
+
+  auto credentials = AwsCredentials{};
+  auto section_name = std::string{};
+  auto line = std::string{};
+  auto profile_found = false;
+
+  while (std::getline(ifs, line)) {
+    auto sline = absl::StripAsciiWhitespace(line);
+    // Ignore empty and commented out lines
+    if (sline.empty() || sline[0] == '#') continue;
+
+    // A configuration section name has been encountered
+    if (sline[0] == '[' && sline[sline.size() - 1] == ']') {
+      section_name =
+          absl::StripAsciiWhitespace(sline.substr(1, sline.size() - 2));
+      continue;
+    }
+
+    // Look for key=value pairs if we're in the appropriate profile
+    if (section_name == profile) {
+      profile_found = true;
+      if (auto pos = sline.find('='); pos != std::string::npos) {
+        auto key = absl::StripAsciiWhitespace(sline.substr(0, pos));
+        auto value = absl::StripAsciiWhitespace(sline.substr(pos + 1));
+
+        if (key == kCfgAwsAccessKeyId) {
+          credentials.access_key = value;
+        } else if (key == kCfgAwsSecretAccessKeyId) {
+          credentials.secret_key = value;
+        } else if (key == kCfgAwsSessionToken) {
+          credentials.session_token = value;
+        }
+      }
+    }
+  }
+
+  if (!profile_found) {
+    return absl::NotFoundError(absl::StrCat("Profile [", profile,
+                                            "] not found "
+                                            "in credentials file [",
+                                            filename, "]"));
+  }
+
+  ABSL_LOG_FIRST_N(INFO, 1)
```

Comment:
See my fancy new logging bits if you want verbose logging:

namespace {
    ABSL_CONST_INIT internal_log::VerboseFlag s3_logging("s3");
}

ABSL_LOG_IF(INFO, s3_logging.Level(1)) << "Using ...";

Reply:
Most unfortunately I didn't have time to look at this in this PR. I've added an item https://github.com/google/tensorstore/issues/114 to review this.